mesonbuild/interpreter/interpreter.py

```diff
@@ -523,6 +523,25 @@ def process_new_values(self, invalues: T.List[T.Union[TYPE_var, ExecutableSerial
             else:
                 raise InterpreterException(f'Module returned a value of unknown type {v!r}.')
 
+    def handle_meson_version(self, pv: str, strict: bool = True) -> None:
```

Comment:
I think the strict arg can be removed, it's always true

Reply:
No, it's needed in the case where the entire AST is correctly built, and we want to validate only the FeatureNew part upfront but wait until after that to error on mismatched versions.

Reply:
```diff
         try:
             self.ast = mparser.Parser(code, mesonfile).parse()
+            self.handle_meson_version_from_ast(False)
```

Reply:
Oh I see, but why waiting if we already know it's going to fail?

Reply:
The difference is in the printed error message. If we raise this directly from the AST, we cannot print the traditional prefix:

```
path/to/meson.build:1:0: ERROR: <errorstring>
```

It is just:
```
ERROR: <errorstring>
```

Reply:
I'm not 100% sure this is something we care about, admittedly.

Reply:
AFAIK you can add `location=kw` into your InterpreterException for that. Every node has the filename, lineno and colno.

Reply:
:thinking: Actually, there's another difference and this one reflects badly on my current approach.

If you use meson_version to guard new functionality inlined into the project() function, we... choke on that *first*, saying:
```
path/to/meson.build:1:0: ERROR: project got unknown keyword arguments "unknownkwarg"
```
And by handling it at the AST level, we could actually warn about the version.

So, it's not that we don't need it, but we can't do it in the end. Thanks for prodding me to think about this, I'll change it.

Reply:
The node trick is awesome and a great point.

Reply:
well.. I guess that's a bit too far, the API we have is MesonException.from_node() but in our case we need to deal with node being None too... forget that :)

Reply:
When is it None? We either enter here via a StringNode, or from func_project() where we can pass self.current_node.

Reply:
Ah indeed, that works too.

Reply:
range-diff:
```diff
1:  f2bdc9dab = 1:  f2bdc9dab Revert "Exit meson with an error if an invalid escape sequence is found in a"
2:  b885a9e09 = 2:  b885a9e09 mparser: use an inherited ParseException everywhere
3:  9cd6e80b7 = 3:  9cd6e80b7 mparser: Add partial AST to exceptions
4:  d5a8ac946 ! 4:  b3b722e93 handle meson_version even when the build file fails to parse
    @@ Commit message
         - teach meson to catch errors in building the initial AST, and just load
           enough of the AST to check for meson_version advice
     
    -    This implements the latter, allowing to future-proofing the build
    +    This implements the latter, allowing to future-proof the build
         grammar.
     
      ## mesonbuild/interpreter/interpreter.py ##
    @@ mesonbuild/interpreter/interpreter.py: class Interpreter(InterpreterBase, Holdab
                  else:
                      raise InterpreterException(f'Module returned a value of unknown type {v!r}.')
      
    -+    def handle_meson_version(self, pv: str, strict: bool = True) -> None:
    -+        if strict and not mesonlib.version_compare(coredata.version, pv):
    -+            raise InterpreterException(f'Meson version is {coredata.version} but project requires {pv}')
    ++    def handle_meson_version(self, pv: str, location: mparser.BaseNode) -> None:
    ++        if not mesonlib.version_compare(coredata.version, pv):
    ++            raise InterpreterException.from_node(f'Meson version is {coredata.version} but project requires {pv}', node=location)
     +
    -+    def handle_meson_version_from_ast(self, strict: bool = True) -> None:
    ++    def handle_meson_version_from_ast(self) -> None:
     +        if not self.ast.lines:
     +            return
     +        project = self.ast.lines[0]
    @@ mesonbuild/interpreter/interpreter.py: class Interpreter(InterpreterBase, Holdab
     +            if kw.value == 'meson_version':
     +                # mypy does not understand "and isinstance"
     +                if isinstance(val, mparser.StringNode):
    -+                    self.handle_meson_version(val.value, strict)
    ++                    self.handle_meson_version(val.value, val)
     +
          def get_build_def_files(self) -> mesonlib.OrderedSet[str]:
              return self.build_def_files
5:  4d1626ec1 ! 5:  1be7e586f interpreter: report FeatureNew for kwargs to project()
    @@ Commit message
     
      ## mesonbuild/interpreter/interpreter.py ##
     @@ mesonbuild/interpreter/interpreter.py: class Interpreter(InterpreterBase, HoldableObject):
    -     def handle_meson_version(self, pv: str, strict: bool = True) -> None:
    -         if strict and not mesonlib.version_compare(coredata.version, pv):
    -             raise InterpreterException(f'Meson version is {coredata.version} but project requires {pv}')
    +     def handle_meson_version(self, pv: str, location: mparser.BaseNode) -> None:
    +         if not mesonlib.version_compare(coredata.version, pv):
    +             raise InterpreterException.from_node(f'Meson version is {coredata.version} but project requires {pv}', node=location)
     +        mesonlib.project_meson_versions[self.subproject] = pv
      
    -     def handle_meson_version_from_ast(self, strict: bool = True) -> None:
    +     def handle_meson_version_from_ast(self) -> None:
              if not self.ast.lines:
     @@ mesonbuild/interpreter/interpreter.py: class Interpreter(InterpreterBase, HoldableObject):
    +         # This needs to be evaluated as early as possible, as meson uses this
              # for things like deprecation testing.
              if kwargs['meson_version']:
    -             self.handle_meson_version(kwargs['meson_version'])
    +-            self.handle_meson_version(kwargs['meson_version'])
     -            mesonlib.project_meson_versions[self.subproject] = kwargs['meson_version']
    ++            self.handle_meson_version(kwargs['meson_version'], node)
      
              if os.path.exists(self.option_file):
                  oi = optinterpreter.OptionInterpreter(self.subproject)
    @@ mesonbuild/interpreterbase/interpreterbase.py: class InterpreterBase:
              assert isinstance(code, str)
              try:
                  self.ast = mparser.Parser(code, mesonfile).parse()
    -+            self.handle_meson_version_from_ast(False)
    ++            self.handle_meson_version_from_ast()
              except mparser.ParseException as me:
                  me.file = mesonfile
                  # try to detect parser errors from new syntax added by future
```